proposals/p6395.md

```diff
@@ -0,0 +1,167 @@
+declarations to be complete, since `extend` changes the scopes which are looked
+in for name lookup.
+
+## Problem
+
+Proposal
+[#5168](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p5168.md)
+laid out rules for when a facet type needs to be identified or complete. When
+`require impls Z` is written, then `Z` must be identified. However it does not
+specify any rule for `extend require impls Z`. or `extend impl as Z`.
+
+If an interface `Y` uses `extend` for another interface `Z`, then a facet type
+for `Y` will also allow name lookup into `Z`. Similarly if a class `Y` uses
+`extend` for an interface `Z`, then an object of type `Y` will also allow name
+lookup into `Z`. In both cases, when the definition of `Y` is complete, we
+should also require `Z` to be complete, as its name scope becomes conceptually a
+part of `Y`.
+
+Additionally, after an `extend ... Z` declaration inside the definition of `Y`,
+the names of `Z` should be immediately available inside the `Y` definition,
+including inside the nested definition of `extend impl as Z { ... }`.
+
+## Background
+
+-   Proposal
+    [#5168](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p5168.md):
+    Forward impl declaration of an incomplete interface
+-   Proposal
+    [#2760](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2760.md):
+    Consistent class and interface syntax
+
+## Proposal
+
+Require that `extend require impls Z` requires `Z` to be complete where the
+declaration is written. Similarly, require that `extend impl as Z` requires `Z`
+to be complete where the declaration is written. In both cases, this ensures
+that `Z` is complete by the time the definition of the enclosing class,
+interface, or named constraint is complete. And it ensures that we can perform
+name lookup into `Z` from within the enclosing scope of the `extend` declaration
+thereafter.
+
+We call a `extend require` declaration _extend-reachable_ from a facet type if
+it's enclosing interface or named constraint is part of that facet type, or if
+its enclosing interface or named constraint is named by another
+_extend-reachable_ declaration. That is, we can find a path from the facet type
+to the `extend require` declaration through other `extend require` declarations.
+
+We say that a facet type is complete if:
+
+-   All its _extend-reachable_ `extend require` declarations are complete. This
+    follows from the enclosing interface or named constraint being complete.
+-   We are able to form a specific for each interface and named constraint
+    extended by the facet type.
+-   We are able to form a specific for each _extend-reachable_ `extend require`
+    declaration in the context of the specific facet type being completed,
+    without monomorphization errors.
+
+## Details
+
+To `extend` an entity `Y` with another `Z` means that name lookups into `Y` will
+also look into `Z`. At the end of the definition of `Y`, we expect to be able to
```

Comment:
Only at the end of the definition of `Y`? I would expect this to apply immediately. For example:
```carbon
base class Z {
   class X {}
}
class Y {
  extend base: Z;
  // Finds `X` in `Z`.
  fn F() -> X;
}
```

Reply:
Yeah, me too. In fact this is meant to come with tests showing that expectation (that got lost in that deleted file mistake). I thought I had added that as an additional step later in the proposal but I don't see it, and there's no need to hedge up here, so yeah let me fix that.